<table border=0 width="100%"><tr><td><p align="left"><img src="Images/logo.png" align="left" width=300></p></td><td><font size=5><B>Python程序设计与应用</B></font></td></tr></table>

# <div align='center'>第七部分 文件和数据格式化</div> 

## 知识点
* 文件、目录、路径的基本概念。
* 从文件中读取数据的办法、语句（或函数、方法）。
* 讲解向文件中写入数据的办法、语句（或函数、方法）。
* 处理CSV 文件和JSON 数据

## 预习
* 第7章：文件和数据格式化

当程序运行时，变量是保存数据的好方法，但如果希望程序结束后数据仍然保持，就需要将数据保存到文件中。另外，在人事管理系统中，你不会希望每次都手动从文件中复制粘贴人事信息到你的代码里。

你可以认为文件的内容是一个字符串值，大小可能有几个GB。

在本部分中，你将学习如何使用Python 在硬盘上创建、读取和保存文件。

# 1. 文件与文件路径
文件是一个存储在辅助存储器上的数据序列，可以包含任何数据内容。概念上，文件是数据的集合和抽象，类似地，函数是程序的集合和抽象。用文件形式组织和表达数据更有效也更为灵活。

文件有两个关键属性：“文件名”（通常写成一个单词）和“路径”。路径指明了文件在计算机上的位置。例如，我的电脑本上有一个文件名为projects.docx，
它的路径在C:\Users\asweigart\Documents。文件名中，最后一个句点之后的部分称为文件的“扩展名”，它指出了文件的类型。project.docx 是一个Word 文档，Users、asweigart和Documents 都是指“文件夹”（也成为目录）。文件夹可以包含文件和其他文件夹。
<img src="Images/file1.png" width=50%>

## 1.1 Windows 上的倒斜杠以及OS X 和Linux 上的正斜杠
在Windows 上，路径书写使用倒斜杠作为文件夹之间的分隔符。但在OS X 和Linux 上，使用正斜杠作为它们的路径分隔符。如果想要程序运行在所有操作系统
上，在编写Python 脚本时，就必须处理这两种情况。

好在，用os.path.join()函数来做这件事很简单。如果将单个文件和路径上的文件夹名称的字符串传递给它，**os.path.join()**就会返回一个文件路径的字符串，包含正确的路径分隔符。

In [6]:
import os
os.path.join('users','swufe','Documents','python.docx')

'users\\swufe\\Documents\\python.docx'

如果需要创建文件名称的字符串，os.path.join()函数就很有用。这些字符串将传递给几个文件相关的函数，本部分将进行介绍。例如，下面的例子将一个文件名列表中的名称，添加到文件夹名称的末尾:

In [7]:
myFiles = ['2019-01-10-1.jpg', 'myPhoto1.png', 'mountain.jpg']
for fileName in myFiles:
    print(os.path.join('d:\\Pictures', fileName))

d:\Pictures\2019-01-10-1.jpg
d:\Pictures\myPhoto1.png
d:\Pictures\mountain.jpg


## 1.2 当前工作目录
每个运行在计算机上的程序，都有一个“当前工作目录”，或cwd。所有没有从根文件夹开始的文件名或路径，都假定在当前工作目录下。利用**os.getcwd()**函数，可以取得当前工作路径的字符串，并可以利用**os.chdir()**改变它。

In [8]:
import os
os.getcwd()

'D:\\SynologyDrive\\学校文档\\教学文档\\程序设计与应用Python\\课件'

In [9]:
os.chdir("c:\\windows\\system32")

In [10]:
os.getcwd()

'c:\\windows\\system32'

## 1.3 绝对路径与相对路径
有两种方法指定一个文件路径。
* “绝对路径”，总是从根文件夹开始。
* “相对路径”，它相对于程序的当前工作目录。

还有点（.）和点点（..）文件夹。它们不是真正的文件夹，而是可以在路径中使用的特殊名称。单个的句点（“点”）用作文件夹目名称时，是“这个目录”的缩写。两个句点（“点点”）意思是父文件夹。

os.path 模块提供了一些函数，返回一个相对路径的绝对路径，以及检查给定的
路径是否为绝对路径。
* 调用**os.path.abspath(path)**将返回参数的绝对路径的字符串。这是将相对路径转换为绝对路径的简便方法。
* 调用**os.path.isabs(path)**，如果参数是一个绝对路径，就返回True，如果参数是一个相对路径，就返回False。
* 调用**os.path.relpath(path, start)**将返回从start 路径到path 的相对路径的字符串。如果没有提供start，就使用当前工作目录作为开始路径。

In [12]:
os.path.abspath('.')

'c:\\windows\\system32'

In [13]:
os.path.abspath('.\\Scripts')

'c:\\windows\\system32\\Scripts'

In [14]:
os.path.isabs('.')

False

In [15]:
os.path.relpath('C:\\Windows', 'C:\\')

'Windows'

In [16]:
os.path.relpath('C:\\Windows', 'C:\\Pictures\\2018')

'..\\..\\Windows'

调用os.path.dirname(path)将返回一个字符串，它包含path 参数中最后一个斜杠之前的所有内容。调用**os.path.basename(path)**将返回一个字符串，它包含path 参数中最后一个斜杠之后的所有内容。

In [17]:
path = 'C:\\Windows\\System32\\calc.exe'

In [18]:
os.path.basename(path)

'calc.exe'

In [19]:
os.path.dirname(path)

'C:\\Windows\\System32'

如果同时需要一个路径的目录名称和基本名称，就可以调用**os.path.split()**，获得这两个字符串的元组，像这样：

In [20]:
calcFilePath = 'C:\\Windows\\System32\\calc.exe'
os.path.split(calcFilePath)

('C:\\Windows\\System32', 'calc.exe')

同时也请注意，os.path.split()不会接受一个文件路径并返回每个文件夹的字符串的列表。如果需要这样，请使用split()字符串方法，并根据os.path.sep 中的字符串进行分割。回忆一下，根据程序运行的计算机，**os.path.sep**变量设置为正确的文件夹分割斜杠。

In [21]:
calcFilePath.split(os.path.sep)

['C:', 'Windows', 'System32', 'calc.exe']

## 1.4 用os.makedirs()创建新文件夹
程序可以用os.makedirs()函数创建新文件夹（目录）。

In [11]:
import os
os.makedirs("c:\\myFolder")

## 1.5 查看文件大小和文件夹内容

一旦有办法处理文件路径，就可以开始搜集特定文件和文件夹的信息。os.path 模块提供了一些函数，用于查看文件的字节数以及给定文件夹中的文件和子文件夹。
* 调用**os.path.getsize(path)**将返回path 参数中文件的字节数。
* 调用os.listdir(path)将返回文件名字符串的列表，包含path 参数中的每个文件（请注意，这个函数在os 模块中，而不是os.path）。

In [22]:
os.path.getsize('C:\\Windows\\System32\\calc.exe')

27648

In [23]:
os.listdir('C:\\Windows\\System32')

['0409',
 '1029',
 '1033',
 '1036',
 '1040',
 '1045',
 '1046',
 '1049',
 '1055',
 '3082',
 '8f629a03-482f-49f2-a27e-279f8764e5cf_HyperV-Compute-Host.dll',
 '@AudioToastIcon.png',
 '@BackgroundAccessToastIcon.png',
 '@bitlockertoastimage.png',
 '@edptoastimage.png',
 '@EnrollmentToastIcon.png',
 '@language_notification_icon.png',
 '@optionalfeatures.png',
 '@VpnToastIcon.png',
 '@WiFiNotificationIcon.png',
 '@windows-hello-V4.1.gif',
 '@WindowsHelloFaceToastIcon.png',
 '@WindowsUpdateToastIcon.contrast-black.png',
 '@WindowsUpdateToastIcon.contrast-white.png',
 '@WindowsUpdateToastIcon.png',
 '@WirelessDisplayToast.png',
 '@WwanNotificationIcon.png',
 '@WwanSimLockIcon.png',
 'aadauthhelper.dll',
 'aadcloudap.dll',
 'aadjcsp.dll',
 'aadtb.dll',
 'aadWamExtension.dll',
 'abcCertFirm.dll',
 'AbcEbankSign.dll',
 'ABCEBankSignCtrl.ocx',
 'abchFTCX22.dll',
 'abchFTCX23.dll',
 'abcsFTCX22.dll',
 'abcsFTCX23.dll',
 'AboutSettingsHandlers.dll',
 'AboveLockAppHost.dll',
 'accessibilitycpl.dll',


如果想知道这个目录下所有文件的总字节数，就可以同时使用os.path.getsize()和os.listdir()。

In [25]:
totalSize = 0
for filename in os.listdir('C:\\Windows\\System32'):
    totalSize = totalSize + os.path.getsize(os.path.join('C:\\Windows\\System32', filename))
print(totalSize/1024/1024)

2375.9645986557007


** 课堂小练习：**
如果想打印出总的文件大小（用GB）表示，应该怎么做？

注意，这不包含System32目录下的文件夹内容的大小。

## 1.6 检查路径有效性
如果你提供的路径不存在，许多Python 函数就会崩溃并报错。os.path 模块提供了一些函数，用于检测给定的路径是否存在，以及它是文件还是文件夹。
* 如果path 参数所指的文件或文件夹存在，调用**os.path.exists(path)**将返回True，否则返回False。
* 如果path 参数存在，并且是一个文件，调用**os.path.isfile(path)**将返回True，否则返回False。
* 如果path 参数存在，并且是一个文件夹，调用**os.path.isdir(path)**将返回True，否则返回False。

In [26]:
os.path.exists('C:\\Windows')

True

In [27]:
os.path.exists('C:\\Windows123456')

False

In [28]:
os.path.isdir('C:\\Windows\\System32')

True

In [29]:
os.path.isfile('C:\\Windows\\System32')

False

In [30]:
os.path.isdir('C:\\Windows\\System32\\calc.exe')

False

In [31]:
os.path.isfile('C:\\Windows\\System32\\calc.exe')

True

# 2.文件读写过程


文件包括两种类型：<br>
* 纯文本文件：“文本文件”只包含基本文本字符，不包含字体、大小和颜色信息。带有.txt 扩展名的文本文件，以及带有.py 扩展名的Python 脚本文件，都是纯文本文件的例子。

* 二进制文件：“二进制文件”是所有其他文件类型，诸如字处理文档、PDF、图像、电子表格和可执行程序。如果用Notepad 或TextEdit 打开一个二进制文件，它看起来就像乱码。二进制文件直接由比特0和比特1组成，没有统一字符编码，文件内部数据的组织格式与文件用途有关。二进制文件和文本文件最主要的区别在于是否有统一的字符编码

无论文件创建为文本文件或者二进制文件，都可以用“文本文件方式”和“二进制文件方式”打开，打开后的操作不同。

In [2]:
# 由于前面操作切换了当前目录，本文件可能放在系统不同的目录结构里，所以这里最好能重启一下内核：点击kernel菜单下的Restart。
import os
textFile = open("Documents/example.txt","rt") #t表示文本文件方式
print(textFile.readline())
textFile.close()
binFile = open("Documents/example.txt","rb")   #b表示二进制文件方式
print(binFile.readline())
binFile.close()

人生苦短，我用Python
b'\xc8\xcb\xc9\xfa\xbf\xe0\xb6\xcc\xa3\xac\xce\xd2\xd3\xc3Python'


在Python 中，读写文件有3 个步骤：<br>
(1) 调用open()函数，返回一个File 对象。<br>
(2) 调用File 对象的read()或write()方法。<br>
(3) 调用File 对象的close()方法，关闭该文件。<br>

## 2.1 用open()函数打开文件
要用**open()**函数打开一个文件，就要向它传递一个字符串路径，表明希望打开的文件。这既可以是绝对路径，也可以是相对路径。open()函数返回一个File对象。

<center>**<变量名> = open(<文件名>, <打开模式>)**</center>

In [ ]:
textFile = open("Documents/example.txt","rt") #t表示文本文件方式

这些命令都将以读取纯文本文件的模式打开文件，或简称为“读模式”。当文件以读模式打开时，Python 只让你从文件中读取数据，你***不能***以任何方式写入或修改它。

在Python 中打开文件时，读模式是默认的模式。但如果你不希望依赖于Python 的默认值，也可以明确指明该模式，向open()传入字符串'r'，作为第二个参数。

调用open()将返回一个File 对象。File 对象代表计算机中的一个文件，它只是Python 中另一种类型的值，就像你已熟悉的列表和字典。

open()函数提供7种基本的打开模式:

|打开模式|含义|
|-----|:-----|
|'r'|只读模式，如果文件不存在，返回异常FileNotFoundError，默认值|
|'w'|覆盖写模式，文件不存在则创建，存在则完全覆盖源文件|
|'x'|创建写模式，文件不存在则创建，存在则返回异常FileExistsError|
|'a'|追加写模式，文件不存在则创建，存在则在原文件最后追加内容|
|'b'|二进制文件模式|
|'t'|文本文件模式，默认值|
|'+'|与r/w/x/a一同使用，在原功能基础上增加同时读写功能|

## 2.2 读取文件内容
既然有了一个File 对象，就可以开始从它读取内容。如果希望将整个文件的内容读取为一个**字符串值**，就使用File对象的read()方法。

In [7]:
empFile = open("Documents/employees.txt")
empInfo = empFile.read()
empFile.close()
print(empInfo)

编号,姓名,部门,电子邮箱,电话号码,身份证号码
10932,张珊,管理,zhans@163.com,18612345678,520125197907167551
10933,李思,软件,lisi@163.com,18612345679,510110195512176250
10934,王武,财务,wangwu@example.com,18612345680,510102196510255826
10935,赵柳,财务,zhaoliu@163.com,18612345681,510115195810287715
10936,钱棋,人事,qianqi@example.com,18612345682,510103196707165650
10941,张明,管理,zhangming@example.com,18612345683,510105198112166015
10942,赵敏,人事,zhaomin@163.com,18612345684,510105198505225815
10945,王红,培训,wanghong@example.com,18612345685,510221196208516555
10946,李萧,培训,lixiao@example.com,18612345686,510221196601510810
10947,孙科,软件,sunke@163.com,18612345687,510105198507025519
10948,刘利,软件,liuli@example.com,18612345688,510227197510280527


或者，可以使用**readlines()**方法，从该文件取得一个字符串的列表。列表中的每个字符串就是文本中的每一行。请注意每行后面的换行符：\\n

In [9]:
empFile = open("Documents/employees.txt")
empInfo = empFile.readlines()
empFile.close()
print(type(empInfo))
print(empInfo)

<class 'list'>
['编号,姓名,部门,电子邮箱,电话号码,身份证号码\n', '10932,张珊,管理,zhans@163.com,18612345678,520125197907167551\n', '10933,李思,软件,lisi@163.com,18612345679,510110195512176250\n', '10934,王武,财务,wangwu@example.com,18612345680,510102196510255826\n', '10935,赵柳,财务,zhaoliu@163.com,18612345681,510115195810287715\n', '10936,钱棋,人事,qianqi@example.com,18612345682,510103196707165650\n', '10941,张明,管理,zhangming@example.com,18612345683,510105198112166015\n', '10942,赵敏,人事,zhaomin@163.com,18612345684,510105198505225815\n', '10945,王红,培训,wanghong@example.com,18612345685,510221196208516555\n', '10946,李萧,培训,lixiao@example.com,18612345686,510221196601510810\n', '10947,孙科,软件,sunke@163.com,18612345687,510105198507025519\n', '10948,刘利,软件,liuli@example.com,18612345688,510227197510280527']


** 课堂练习 **

请统计employees.txt文件的中女员工的个数：

首先定义一个函数，根据身份证号码判断是不是女员工（是女员工返回1，不是返回0）：

In [10]:
def isFemaleByIDCard(idCard):
    if len(idCard) == 18:
        return 1 if int(idCard[16]) % 2 == 0 else 0  #理解一下这句话 :)
    elif len(idCard) == 15:
        return 1 if int(idCard[13] % 2) == 0 else 0
    else:
        return 0

第2步，将employees.txt文件的内容读入，并存储到empList中：

In [11]:
empFile = open("Documents/employees.txt")
empInfo = empFile.readlines()
empFile.close()

empList = []
for i in range(1,len(empInfo)):  #注意，这里是从1开始的，为什么？
    empStr = empInfo[i]
    empStr = empStr.strip('\n')  # 这里是做什么？
    empList.append(empStr.split(','))  #这里又是在做什么？
print(empList)

[['10932', '张珊', '管理', 'zhans@163.com', '18612345678', '520125197907167551'], ['10933', '李思', '软件', 'lisi@163.com', '18612345679', '510110195512176250'], ['10934', '王武', '财务', 'wangwu@example.com', '18612345680', '510102196510255826'], ['10935', '赵柳', '财务', 'zhaoliu@163.com', '18612345681', '510115195810287715'], ['10936', '钱棋', '人事', 'qianqi@example.com', '18612345682', '510103196707165650'], ['10941', '张明', '管理', 'zhangming@example.com', '18612345683', '510105198112166015'], ['10942', '赵敏', '人事', 'zhaomin@163.com', '18612345684', '510105198505225815'], ['10945', '王红', '培训', 'wanghong@example.com', '18612345685', '510221196208516555'], ['10946', '李萧', '培训', 'lixiao@example.com', '18612345686', '510221196601510810'], ['10947', '孙科', '软件', 'sunke@163.com', '18612345687', '510105198507025519'], ['10948', '刘利', '软件', 'liuli@example.com', '18612345688', '510227197510280527']]


第3步，调用isFemaleByIDCard函数进行统计：

In [12]:
femaleCount = 0
for emp in empList:
    femaleCount += isFemaleByIDCard(emp[5])

print("女员工的人数是：", femaleCount)

女员工的人数是： 2


根据打开方式不同可以对文件进行相应的读写操作，Python提供4个常用的文件内容读取方法：

|方法|含义|
|---|:---|
|&lt;file&gt;.readall()|读入整个文件内容，返回一个字符串或字节流|
|&lt;file&gt;.read(size=-1)|从文件中读入整个文件内容，如果给出参数，读入前size长度的字符串或字节流|
|&lt;file&gt;.readline(size = -1)|从文件中读入一行内容，如果给出参数，读入该行前size长度的字符串或字节流|
|&lt;file&gt;.readlines(hint=-1)|从文件中读入所有行，以每行为元素形成一个列表，如果给出参数，读入hint行|


文本文件的逐行打印：

In [25]:
with open("Documents/employees.txt","r") as empFile:
    for line in empFile:
        print(line)

编号,姓名,部门,电子邮箱,电话号码,身份证号码

10932,张珊,管理,zhans@163.com,18612345678,520125197907167551

10933,李思,软件,lisi@163.com,18612345679,510110195512176250

10934,王武,财务,wangwu@example.com,18612345680,510102196510255826

10935,赵柳,财务,zhaoliu@163.com,18612345681,510115195810287715

10936,钱棋,人事,qianqi@example.com,18612345682,510103196707165650

10941,张明,管理,zhangming@example.com,18612345683,51010519811216015

10942,赵敏,人事,zhaomin@163.com,18612345684,510105198505225815

10945,王红,培训,wanghong@example.com,18612345685,510221196208516555

10946,李萧,培训,lixiao@example.com,18612345686,510221196601510810

10947,孙科,软件,sunke@163.com,18612345687,510105198507025519

10948,刘利,软件,liuli@example.com,18612345688,510227197510280527


上面的方法用到了with语句，使用with语句，可以不需要调用文件的close()方法，文件会在with语句结束后自动关闭。

如果程序需要逐行处理文件内容，建议采用上述代码格式。特别是对于大型文件的读取，这样处理比较合理。

## 2.3 写入文件
Python 允许你将内容写入文件，方式与print()函数将字符串“写”到屏幕上类似。但是，如果打开文件时用读模式，就不能写入文件。你需要以“写入纯文本模
式”或“添加纯文本模式”打开该文件，或简称为“写模式”和“添加模式”。写模式将覆写原有的文件，从头开始，就像你用一个新值覆写一个变量的值。

将'**w**'作为第二个参数传递给open()，以写模式打开该文件。不同的是，添加模式将在已有文件的末尾添加文本。可以认为这类似向一个变量中的列表添加内容，而不是完全覆写该变量。

将'**a**'作为第二个参数传递给open()，以添加模式打开该文件。

如果传递给 open()的文件名不存在，写模式和添加模式都会创建一个新的空文件。在读取或写入文件后，调用close()方法，然后才能再次打开该文件。

**课堂练习**

接着上面的例子，将女员工人数，男员工人数和身份证号码有误的员工写入到统计结果文件result.txt:

类似的，定义一个函数，根据身份证号码判断是不是男员工（是女员工返回1，不是返回0）：

In [28]:
def isMaleByIDCard(idCard):
    if len(idCard) == 18:
        return 1 if int(idCard[16]) % 2 == 1 else 0  #理解一下这句话 :)
    elif len(idCard) == 15:
        return 1 if int(idCard[13] % 2) == 1 else 0
    else:
        return 0

再定义一个函数，判断传入的身份证号码是否有误（身份证号码应该是18位或者15位），如果号码格式正确，返回True；如果格式不正确，返回False：

In [29]:
def isCorrectIDCard(idCard):
    if len(idCard) in (15, 18):
        return True
    else:
        return False

***PS. 这个方法判断身份证号码的正确性是简单粗暴的，如果要实现更复杂的判断规则，也许你想了解一下<font color="red">正则表达式</font>，这会是扩展的内容。***

In [30]:
femaleCount = 0
maleCount = 0
invalidIDCard = []

empFile = open("Documents/employees.txt")
empInfo = empFile.readlines()
empFile.close()

empList = []
for i in range(1,len(empInfo)): 
    empStr = empInfo[i]
    empStr = empStr.strip('\n')  
    empList.append(empStr.split(',')) 

for emp in empList:
    femaleCount += isFemaleByIDCard(emp[5])
    maleCount += isMaleByIDCard(emp[5])
    if isCorrectIDCard(emp[5]) == False:
        invalidIDCard.append(emp[1])

print("总人数：", len(empList))
print("女员工人数：", femaleCount)
print("男员工人数：", maleCount)
if len(invalidIDCard)>0:
    print("以下员工身份证号码有误：", end="")
    for name in invalidIDCard:
        print(name, end=" ")

总人数： 11
女员工人数： 2
男员工人数： 7
以下员工身份证号码有误：李思 张明 

将统计的内容写入到result.txt文件：

In [31]:
resultFile = open("Documents/result.txt","w")
resultFile.write("总人数：{}\n".format(len(empList)))
resultFile.write("女员工人数：{}\n".format(femaleCount))
resultFile.write("男员工人数：{}\n".format(maleCount))
if len(invalidIDCard) > 0:
    nameStr = ""
    for name in invalidIDCard:
        nameStr += " " + name

resultFile.write("\n")
resultFile.write("有{}个员工身份证号码有误：{}\n".format(len(invalidIDCard),nameStr))
resultFile.close()

想想你们有500个员工，这个统计效率是不是很高了？

Python提供3个与文件内容写入有关的方法:

|方法|含义|
|---|:---|
|&lt;file&gt;.write(s)|向文件写入一个字符串或字节流|
|&lt;file&gt;.writelines(lines)|将一个元素为字符串的列表写入文件|
|&lt;file&gt;.seek(offset)|改变当前文件操作指针的位置，offset的值：0：文件开头； 1: 当前位置； 2: 文件结尾|

上面的统计内容，可以使用wirtelines()函数将身份证号码有误的员工姓名一次写入文件：

In [32]:
with open("Documents/result.txt","w") as resultFile:
    resultFile.write("总人数：{}\n".format(len(empList)))
    resultFile.write("女员工人数：{}\n".format(femaleCount))
    resultFile.write("男员工人数：{}\n".format(maleCount))
    if len(invalidIDCard) > 0:
        resultFile.write("\n")
        resultFile.write("有{}个员工身份证号码有误：\n".format(len(invalidIDCard)))
        
        resultFile.writelines(invalidIDCard)

看看下面的结果，好像有点问题：

<img src="Images/file2.png" width=50%>

解决这个问题的方法会有很多，其中比较方便的一种是列表推导式（非常Python的一种写法，也许你想了解以下）：

In [33]:
with open("Documents/result.txt","w") as resultFile:
    resultFile.write("总人数：{}\n".format(len(empList)))
    resultFile.write("女员工人数：{}\n".format(femaleCount))
    resultFile.write("男员工人数：{}\n".format(maleCount))
    if len(invalidIDCard) > 0:
        resultFile.write("\n")
        resultFile.write("有{}个员工身份证号码有误：\n".format(len(invalidIDCard)))
        
        resultFile.writelines([item + " " for item in invalidIDCard])

# 3.处理CSV文件

CSV 表示“Comma-Separated Values（逗号分隔的值）”，CSV文件是简化的电子表格，保存为纯文本文件。它是一种通用的、相对简单的文件格式，在商业和科学上广泛应用，尤其应用在程序之间转移表格数据。

上面处理的员工信息格式其实就是一种CSV文件的格式。该格式的应用有一些基本规则，如下：

（1）纯文本格式，通过单一编码表示字符；<br>
（2）以行为单位，开头不留空行，行之间没有空行；<br>
（3）每行表示一个一维数据，多行表示二维数据；<br>
（4）以逗号分隔每列数据，列数据为空也要保留逗号；<br>
（5）可以包含或不包含列名，包含时列名放置在文件第一行。<br>


Python 的csv模块让解析CSV文件变得容易。

## 3.1 csv 模块
CSV 文件中的每行代表电子表格中的一行，逗号分割了该行中的单元格。在一个CSV 文件中，看起来像这样：

<img src="Images/file3.png" width=40%>

CSV 文件是简单的，缺少Excel 电子表格的许多功能。例如，CSV 文件中：
* 值没有类型，所有东西都是字符串；
* 没有字体大小或颜色的设置；
* 没有多个工作表；
* 不能指定单元格的宽度和高度；
* 不能合并单元格；
* 不能嵌入图像或图表。

CSV 的文件的优势是简单。

csv模块是Python 自带的，所以不需要安装就可以导入它。

## 3.2 Reader 对象
要用csv模块从CSV 文件中读取数据，需要创建一个Reader 对象。Reader 对象让你迭代遍历CSV 文件中的每一行。

In [34]:
import csv

incomeFile = open('Documents/income.csv')
incomeReader = csv.reader(incomeFile)
incomeList = list(incomeReader)
incomeList

[['编号', '姓名', '基本工资', '奖金', '住房补助', '五险一金', '其他扣款', '专项附加扣除'],
 ['10932',
  '张珊',
  '15000.00 ',
  '4000.00 ',
  '2300.00 ',
  '980.00 ',
  '300.00 ',
  '1000.00'],
 ['10933',
  '李思',
  '12000.00 ',
  '5000.00 ',
  '2600.00 ',
  '890.00 ',
  '400.00 ',
  '2000.00'],
 ['10934',
  '王武',
  '10100.00 ',
  '2000.00 ',
  '2500.00 ',
  '860.00 ',
  '500.00 ',
  '1500.00'],
 ['10935',
  '赵柳',
  '10050.00 ',
  '1000.00 ',
  '2700.00 ',
  '1080.00 ',
  '300.00 ',
  '2000.00'],
 ['10936',
  '钱棋',
  '10200.00 ',
  '2000.00 ',
  '2400.00 ',
  '1000.00 ',
  '600.00 ',
  '1000.00'],
 ['10941',
  '张明',
  '13600.00 ',
  '4000.00 ',
  '2100.00 ',
  '680.00 ',
  '300.00 ',
  '1000.00'],
 ['10942',
  '赵敏',
  '13200.00 ',
  '2500.00 ',
  '2300.00 ',
  '670.00 ',
  '400.00 ',
  '3000.00'],
 ['10945',
  '王红',
  '13600.00 ',
  '2600.00 ',
  '2030.00 ',
  '560.00 ',
  '400.00 ',
  '2000.00'],
 ['10946',
  '李萧',
  '12500.00 ',
  '2800.00 ',
  '2400.00 ',
  '420.00 ',
  '500.00 ',
  '1500.00'],
 ['10947',
  '孙科'

回头和前面对于员工信息处理的方法比较一下，是不是更简单了？

既然已经将CSV文件表示为列表的列表，就可以用表达式**incomeList[row][col]**来访问特定行和列的值。其中，row是incomeList中一个列表的下标，col是该列表中你想访问的项的下标。

In [36]:
incomeList[1][1]

'张珊'

In [38]:
incomeList[0][5]

'五险一金'

In [39]:
incomeList[1][5]

'980.00 '

## 3.3 在for 循环中，从Reader 对象读取数据
对于大型的CSV 文件，你需要在一个for 循环中使用Reader 对象。这样避免将整个文件一次性装入内存。例如：

In [40]:
import csv

incomeFile = open('Documents/income.csv')
incomeReader = csv.reader(incomeFile)

for row in incomeReader:
    print(" ".join(row))

编号 姓名 基本工资 奖金 住房补助 五险一金 其他扣款 专项附加扣除
10932 张珊 15000.00  4000.00  2300.00  980.00  300.00  1000.00
10933 李思 12000.00  5000.00  2600.00  890.00  400.00  2000.00
10934 王武 10100.00  2000.00  2500.00  860.00  500.00  1500.00
10935 赵柳 10050.00  1000.00  2700.00  1080.00  300.00  2000.00
10936 钱棋 10200.00  2000.00  2400.00  1000.00  600.00  1000.00
10941 张明 13600.00  4000.00  2100.00  680.00  300.00  1000.00
10942 赵敏 13200.00  2500.00  2300.00  670.00  400.00  3000.00
10945 王红 13600.00  2600.00  2030.00  560.00  400.00  2000.00
10946 李萧 12500.00  2800.00  2400.00  420.00  500.00  1500.00
10947 孙科 12000.00  3500.00  2030.00  880.00  300.00  1000.00
10948 刘利 14200.00  2500.00  2020.00  900.00  400.00  2000.00


Reader对象只能循环遍历一次。要再次读取CSV文件，必须调用csv.reader，创建一个对象。例如，下面的代码什么都打印不出来：

In [42]:
for row in incomeReader:
    print(" ".join(row))

## 3.4 Writer 对象
利用Writer对象可以将数据写 CSV文件。要创建一个 Writer对象，就使用csv.writer()函数。

将income.csv文件中的每个人的基本工资、奖金和住房补助加起来成为收入；五险一金、其他扣款和专项附加扣除加起来成为扣除，并将新的数据写入到文件incomeTotal.csv文件中：

In [52]:
import csv

incomeFile = open('Documents/income.csv')
incomeReader = csv.reader(incomeFile)

incomeTotalRows = []
incomeTotalRows.append(['编号', '姓名', '收入', '扣除'])

for row in incomeReader:
    if incomeReader.line_num == 1:  #这句是做什么？
        continue

    incomeTotalRows.append([row[0], row[1], float(row[2]) +float(row[3]) + float(row[4]), float(row[5]) + float(row[6]) + float(row[7])])

incomeFile.close()

for row in incomeTotalRows:
    print(" ".join([str(item) for item in row])) #这里又用到了列表推导式

编号 姓名 收入 扣除
10932 张珊 21300.0 2280.0
10933 李思 19600.0 3290.0
10934 王武 14600.0 2860.0
10935 赵柳 13750.0 3380.0
10936 钱棋 14600.0 2600.0
10941 张明 19700.0 1980.0
10942 赵敏 18000.0 4070.0
10945 王红 18230.0 2960.0
10946 李萧 17700.0 2420.0
10947 孙科 17530.0 2180.0
10948 刘利 18720.0 3300.0


写入CSV 文件:

In [53]:
incomeTotalFile = open('Documents/incomeTotal.csv', 'w', newline="")  #newline什么作用？上网查一查
incomeTotalCSVWriter = csv.writer(incomeTotalFile)

for row in incomeTotalRows:
    incomeTotalCSVWriter.writerow(row)
    
incomeTotalFile.close()

# 4. 高维数据的格式化JSON

JavaScript对象表示法是一种流行的方式，将数据格式化，成为人可读的字符串。JSON是JavaScript 程序编写数据结构的原生方式。不需要了解JavaScript，也能处理JSON格式的数据。

与一维二维数据不同，高维数据能展示数据间更为复杂的组织关系。为了保持灵活性，表示高维数据不采用任何结构形式，仅采用最基本的二元关系，即键值对。万维网是高维数据最成功的典型应用。
JSON格式可以对高维数据进行表达和存储。JSON（JavaScript Object Notation）是一种轻量级的数据交换格式，易于阅读和理解。JSON格式表达键值对<key, value>的基本格式如下，键值对都保存在双引号中：
<center> ** "key" : "value" ** </center>

当多个键值对放在一起时，JSON有如下一些约定：
* 数据保存在键值对中；
* 键值对之间由逗号分隔；
* 括号用于保存键值对数据组成的对象；
* 方括号用于保存键值对数据组成的数组。

下面是JSON 格式数据的一个例子：

```Python
{
"empNo":"10932",
 "empName":"张珊",
 "incomeSum":21300.00,
 "deductionSum":2280.00
 }
```
是不是和字典很像？注意，这是一个JSON字符串。

了解JSON 是很有用，因为很多网站都提供JSON 格式的内容，作为程序与网站交互的方式。这就是所谓的提供“应用程序编程接口（API）”。访问API 和通过
URL 访问任何其他网页是一样的。不同的是，API 返回的数据是针对机器格式化的（例如用JSON），API 不是人容易阅读的。

## 4.1 json 模块
Python 的json 模块处理了JSON 数据字符串和Python 值之间转换的所有细节，得到了**json.loads()**和**json.dumps()**函数。

JSON 不能存储每一种Python 值，它只能包含以下数据类型的值：字符串、整型、浮点型、布尔型、列表、字典和NoneType。

JSON 不能表示Python 特有的对象，如File 对象、CSV Reader 或Writer 对象、Regex对象或Selenium WebElement 对象。

### 4.1.1 用loads()函数读取JSON
要将包含JSON 数据的字符串转换为Python 的值，就将它传递给json.loads()函数（这个名字的意思是“load string”，而不是“loads”）：

In [55]:
empJsonData = '{"empNo":"10932","empName":"张珊","incomeSum":21300.00,"deductionSum":2280.00}'

import json
empPythonValue = json.loads(empJsonData)
print(type(empPythonValue))
print()
print(empPythonValue)

<class 'dict'>

{'empNo': '10932', 'empName': '张珊', 'incomeSum': 21300.0, 'deductionSum': 2280.0}


In [57]:
moreEmpJsonData = '[{"empNo":"10932","empName":"张珊","incomeSum":21300.00,"deductionSum":2280.00},\
{"empNo":"10933","empName":"李思","incomeSum":19600.00,"deductionSum":3290.00},\
{"empNo":"10934","empName":"王武","incomeSum":14600.00,"deductionSum":2860.00}]'

import json
moreEmpPythonValue = json.loads(moreEmpJsonData)
print(type(moreEmpPythonValue))
print()
print(moreEmpPythonValue)

<class 'list'>

[{'empNo': '10932', 'empName': '张珊', 'incomeSum': 21300.0, 'deductionSum': 2280.0}, {'empNo': '10933', 'empName': '李思', 'incomeSum': 19600.0, 'deductionSum': 3290.0}, {'empNo': '10934', 'empName': '王武', 'incomeSum': 14600.0, 'deductionSum': 2860.0}]


### 4.1.2 用dumps 函数写出JSON
json.dumps()函数（它表示“dump string”，而不是“dumps”）将一个Python 值转换成JSON 格式的数据字符串。

In [59]:
empDicValue = {"empNo":"10932","empName":"张珊","incomeSum":21300.00,"deductionSum":2280.00}  #这是一个Python字典
print(type(empDicValue))

<class 'dict'>


In [60]:
empJsonStr = json.dumps(empDicValue)  #empJsonStr是一个字符串
print(type(empJsonStr))
print()
print(empJsonStr)

<class 'str'>

{"empNo": "10932", "empName": "\u5f20\u73ca", "incomeSum": 21300.0, "deductionSum": 2280.0}


有了Json，就可以将Python中的列表，字典等对象转换成字符串，在互联网上传输；也可以从互联网获取Json字符串，转换成Python中的字典或列表。

## 4.2 实例：豆瓣电影数据分析

In [85]:
import json, requests

url = 'http://api.douban.com/v2/movie/top250'
response = requests.get(url)  # 这里可能需要你更进一步学习Python
start=1
count=25
response = requests.get(url, params={'start': start, 'count': count})
response.encoding='UTF_8'
content=response.json()

下面来分析一下这个content:

In [83]:
print(type(content))
print(content)

<class 'dict'>
{'count': 25, 'start': 1, 'total': 250, 'subjects': [{'rating': {'max': 10, 'average': 9.6, 'stars': '50', 'min': 0}, 'genres': ['剧情', '爱情', '同性'], 'title': '霸王别姬', 'casts': [{'alt': 'https://movie.douban.com/celebrity/1003494/', 'avatars': {'small': 'http://img3.doubanio.com/view/celebrity/s_ratio_celebrity/public/p67.jpg', 'large': 'http://img3.doubanio.com/view/celebrity/s_ratio_celebrity/public/p67.jpg', 'medium': 'http://img3.doubanio.com/view/celebrity/s_ratio_celebrity/public/p67.jpg'}, 'name': '张国荣', 'id': '1003494'}, {'alt': 'https://movie.douban.com/celebrity/1050265/', 'avatars': {'small': 'http://img1.doubanio.com/view/celebrity/s_ratio_celebrity/public/p46345.jpg', 'large': 'http://img1.doubanio.com/view/celebrity/s_ratio_celebrity/public/p46345.jpg', 'medium': 'http://img1.doubanio.com/view/celebrity/s_ratio_celebrity/public/p46345.jpg'}, 'name': '张丰毅', 'id': '1050265'}, {'alt': 'https://movie.douban.com/celebrity/1035641/', 'avatars': {'small': 'http://img

content是个字典，包括了count, start, total和subjects这些键，他们分别可能是什么？

看看subjects里面的东西：

In [84]:
movies = content['subjects']

In [129]:
movies[3]

{'alt': 'https://movie.douban.com/subject/1292063/',
 'casts': [{'alt': 'https://movie.douban.com/celebrity/1041004/',
   'avatars': {'large': 'http://img1.doubanio.com/view/celebrity/s_ratio_celebrity/public/p26764.jpg',
    'medium': 'http://img1.doubanio.com/view/celebrity/s_ratio_celebrity/public/p26764.jpg',
    'small': 'http://img1.doubanio.com/view/celebrity/s_ratio_celebrity/public/p26764.jpg'},
   'id': '1041004',
   'name': '罗伯托·贝尼尼'},
  {'alt': 'https://movie.douban.com/celebrity/1000375/',
   'avatars': {'large': 'http://img3.doubanio.com/view/celebrity/s_ratio_celebrity/public/p9548.jpg',
    'medium': 'http://img3.doubanio.com/view/celebrity/s_ratio_celebrity/public/p9548.jpg',
    'small': 'http://img3.doubanio.com/view/celebrity/s_ratio_celebrity/public/p9548.jpg'},
   'id': '1000375',
   'name': '尼可莱塔·布拉斯基'},
  {'alt': 'https://movie.douban.com/celebrity/1000368/',
   'avatars': {'large': 'http://img1.doubanio.com/view/celebrity/s_ratio_celebrity/public/p45590.jpg',
 

看上去是一部电影的信息。 

打印出每部电影的名称，评分，年份，主演，导演：

In [136]:
for movie in movies:
    title = movie['title'].strip()
    year = movie['year'].strip()
    rate = movie['rating']['average']
    casts = ','.join([item['name'] for item in movie['casts']])
    directors = ','.join([item['name'] for item in movie['directors']])
    
    print('{title:<{len1}}{year:<6}{rate:3.1f}{casts:<{len2}}{directors:<{len3}}'.format(title=title,len1=22-len(title.encode('GBK'))+len(title), 
                                                                           year=year, 
                                                                           rate=rate, 
                                                                           casts=casts,len2=55-len(casts.encode('GBK'))+len(casts), 
                                                                           directors=directors,len3=22-len(directors.encode('GBK'))+len(directors)))


霸王别姬              1993  9.6张国荣,张丰毅,巩俐                                     陈凯歌                
这个杀手不太冷        1994  9.4让·雷诺,娜塔莉·波特曼,加里·奥德曼                   吕克·贝松            
阿甘正传              1994  9.4汤姆·汉克斯,罗宾·怀特,加里·西尼斯                   罗伯特·泽米吉斯      
美丽人生              1997  9.5罗伯托·贝尼尼,尼可莱塔·布拉斯基,乔治·坎塔里尼       罗伯托·贝尼尼        
泰坦尼克号            1997  9.3莱昂纳多·迪卡普里奥,凯特·温丝莱特,比利·赞恩         詹姆斯·卡梅隆        
千与千寻              2001  9.3柊瑠美,入野自由,夏木真理                               宫崎骏                
辛德勒的名单          1993  9.5连姆·尼森,本·金斯利,拉尔夫·费因斯                   史蒂文·斯皮尔伯格    
盗梦空间              2010  9.3莱昂纳多·迪卡普里奥,约瑟夫·高登-莱维特,艾伦·佩吉    克里斯托弗·诺兰      
忠犬八公的故事        2009  9.3理查·基尔,萨拉·罗默尔,琼·艾伦                       拉斯·霍尔斯道姆      
机器人总动员          2008  9.3本·贝尔特,艾丽莎·奈特,杰夫·格尔林                   安德鲁·斯坦顿        
三傻大闹宝莱坞        2009  9.2阿米尔·汗,卡琳娜·卡普尔,马达范                       拉吉库马尔·希拉尼    
海上钢琴师            1998  9.2蒂姆·罗斯,普路特·泰勒·文斯,比尔·努恩               朱塞佩·托纳多雷      
放牛班的春天          2004  9.3热拉尔·朱尼奥,弗朗索瓦·贝莱昂,

*** 上面的format语句现在理解不到没有关系，也许你可以在网上查一查。***

### 思考：如果希望按照评分由高到低打印出来，可以怎么做？

# 5. 小结

文件被组织在文件夹中（也称为目录），路径描述了一个文件的位置。运行在计算机上的每个程序都有一个当前工作目录，它让你相对于当前的位置指定文件路径，而非总是需要完整路径（绝对路径）。os.path 模块包含许多函数，用于操作文件路径。

程序也可以直接操作文本文件的内容。open()函数将打开这些文件，将它们的内容读取为一个大字符串（利用read()方法），或读取为字符串的列表（利用方
法readlines()）。Open()函数可以将文件以写模式或添加模式打开，分别创建新的文本文件或在原有的文本文件中添加内容。


CSV 和JSON 是常见的纯文本格式，用于保存数据。它们很容易被程序解析，同时仍然让人可读，所以它们经常被用作简单的电子表格或网络应用程序的数据。csv 和json 模块大大简化了读取和写入CSV 和JSON 文件的过程。

一个常见的任务是接受多种格式的数据，解析它，并获得需要的特定信息。这些任务往往非常特别，商业软件并不是最有帮助的。通过编写自己的脚本，可以让计算机处理大量以这些格式呈现的数据。

## 练习
**(1)** 相对路径是相对于什么？<br>

**(2)** 绝对路径从什么开始？<br>

**(3)** os.getcwd()和os.chdir()函数做什么事？<br>

**(4)** read()和readlines()方法之间的区别是什么？

**(5)** 向csv.reader()和csv.writer()传入什么，来创建Reader 和Writer 对象？

**(6)** 对于Reader 和Writer 对象，File 对象需要以什么模式打开？<br>

**(7)** 什么函数接受一个JSON 数据的字符串，并返回一个Python 数据结构？<br>

**(8)** 什么函数接受一个Python 数据结构，并返回一个JSON 数据的字符串？<br>

# 作业：商城销售统计

假设你是学生实验超市的销售总监，每个月你会收到销售数据，这些数据存在CSV文件中，文件内容看起来像这样：
<img src="Images/homework71.png" width="70%">

其中order_id是订单号，goods_id是商品代码，goods_name是商品名称，unit_price是商品单价，quantity是销售数量，return_qty是退货数量。

现在你有2018年1月至4月份销售数据，分别存在2018_01.csv，2018_02.csv，2018_03.csv，2018_04.csv文件中，请做出以下统计：
* 每个月的销售额，以及4个月的销售总额写入到salesAmount.txt文件中，每个月的数据和总额数据各占1行（共5行）。你的文件应该看起来像：
<img src="Images/homework71_2.png" width=30%>
* 每个月销售额最高的10种商品的商品代码，商品名称，销售数量，销售额分别写入到1个文件中，文件名分别为：salesTop10_01.txt，salesTop10_02.txt，salesTop10_03.txt，salesTop10_04.txt。你的每个文件应该看起来像：
<img src="Images/homework71_3.png" width=30%>

将作业的程序源文件保存在12345678.py文件中作为第5次作业的附件,其中12345678需要替换成你的学号,发送到xzl_swufe@163.com。<br><br>
<b><font color=red>邮件标题必须类似于：作业7_12345678。</font>其中12345678需要替换成你的学号。<br><br><b><font color=red>邮件标题必须类似于：作业7_12345678。</font>其中12345678需要替换成你的学号。<br><br><b><font color=red>邮件标题必须类似于：作业7_12345678。</font>其中12345678需要替换成你的学号。